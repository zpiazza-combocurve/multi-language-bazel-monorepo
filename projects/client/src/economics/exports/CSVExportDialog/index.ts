import CustomReportDialogTheme from './theme';

export * from './helpers';
export * from './types';
export * from './Tour';

export { CustomReportDialogTheme };
