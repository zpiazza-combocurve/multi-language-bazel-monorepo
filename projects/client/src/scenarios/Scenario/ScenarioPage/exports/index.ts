export { ExportToAriesDialog } from './ExportToAriesDialog';
export { ExportToCsvDialog } from './ExportToCsvDialog';
export { ExportToPhdwinDialog } from './ExportToPhdwinDialog';
export * from './ExportProbabilisticDialog';
export { default as CCImportDialog } from './CCImportDialog';
export * from './CCExportDialog';
export * from './useCCExport';
