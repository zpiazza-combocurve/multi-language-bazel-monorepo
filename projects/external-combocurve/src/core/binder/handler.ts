/* eslint-disable complexity, @typescript-eslint/ban-types */
import 'reflect-metadata';
import { Request } from 'express';

import { RequestStructureError, ValidationError } from '@src/helpers/validation';

import { FieldMetadata, MetadaWrapper } from '../metadata/metada-wrapper';
import { IInternalRequestBindOptions, NamingTypes, ParseOutput } from '../common';

import { appParses } from './parses';
import { RequestStructure } from './request-structure';
import { sourceFieldLookup } from './field-lookup';

const propertyNotFound = Symbol('propertyNotFound');

// Just a helper type to avoid write Record<string, unknown> everywhere
type JsObj = Record<string, unknown>;

export class BindHandler<TRequestType extends {}> {
	errors: ValidationError[] = [];

	constructor(
		private requestModel: TRequestType,
		private request: Request,
		private metadata: MetadaWrapper,
		private requestStructure: RequestStructure,
		private namingConfig = NamingTypes.ExactlyEqual,
	) {}

	public getModel(): TRequestType {
		return this.requestModel;
	}

	public bind(): ValidationError[] {
		this.innerBind('', this.requestModel, this.metadata.fields);
		return this.errors;
	}

	private innerBind(
		location: string,
		modelAux: JsObj,
		fields: FieldMetadata[],
		scopedSource: unknown = undefined,
	): void {
		// Check unknown fields on source
		if (scopedSource !== undefined) {
			this.errors.push(
				...this.requestStructure.checkUnknowInfoOnSource(
					scopedSource as Record<string, unknown>,
					fields.map((f) => f.name),
					location,
				),
			);
		}

		for (const field of fields) {
			if (field.options.fromWhere === 'composition' && field.innerFields) {
				if (modelAux[field.name] === undefined) {
					modelAux[field.name] = getInstance(field.options.objFactory);
				}

				const innerModel = modelAux[field.name] as JsObj;
				this.innerBind(location, innerModel, field.innerFields.fields, undefined);

				continue;
			}

			const requestValue = this.getPropertyFromRequest(field, scopedSource);

			// If the property is not found, it's not a problem if the property is optional
			if (requestValue === propertyNotFound) {
				if (field.options.isOptional !== true) {
					this.errors.push(new RequestStructureError(`Property ${field.name} not found`, location));
				}

				continue;
			}

			// When object is expected, It's create a scoped that is the object value from request
			// And then called this same bind recursively to fill the object using this new scope
			if (field.options.expects === 'object' && field.options.customParse === undefined && field.innerFields) {
				if (typeof requestValue !== 'object') {
					this.errors.push(new RequestStructureError(`Property ${field.name} is not an object`, location));
				}

				modelAux[field.name] = getInstance(field.options.objFactory);

				this.innerBind(
					location + field.name,
					modelAux[field.name] as JsObj,
					field.innerFields.fields,
					requestValue,
				);

				continue;
			}

			let parsedValue = requestValue;

			// We don't need to parse services since they are come from our context
			if (field.options.expects !== 'service') {
				parsedValue = this.getPropertParsedValue(field, requestValue, location);
			}

			modelAux[field.name] = parsedValue;
		}
	}

	private getPropertyFromRequest(field: FieldMetadata, scopedSource: unknown = undefined): unknown {
		const source = this.requestStructure.getRequestSource(field.options.fromWhere, scopedSource);

		// We accept both array and single object for body
		if (field.options.fromWhere === 'body' && field.options.isBody) {
			return field.options.expects === 'array' && !Array.isArray(source) ? [source] : source;
		}

		const keyValue = sourceFieldLookup(source, field.name, this.namingConfig);
		return keyValue?.value ?? propertyNotFound;
	}

	private getPropertParsedValue(
		field: FieldMetadata,
		requestValue: unknown,
		location: string,
		overrideExpects?: string,
	): unknown {
		const expectType = overrideExpects ?? field.options.expects;

		if (expectType === 'array') {
			return this.getArrayPropertParsedValue(field, requestValue, location);
		}

		let parseFN = appParses[expectType];
		if (field.options.customParse !== undefined) {
			parseFN = field.options.customParse;
		}

		const output = parseFN({
			name: field.name,
			value: requestValue,
			location,
			model: this.requestModel,
			request: this.request,
			requirements: field.options.requirements,
		});
		if (output.errorMsg) {
			this.errors.push(
				output.errorName
					? new ValidationError(output.errorMsg, location, output.errorName)
					: new RequestStructureError(output.errorMsg, location),
			);
			return;
		}

		this.checkTargetConstructor(field.options, output, location);
		return output.parsedValue;
	}

	// Fill Array
	private getArrayPropertParsedValue(field: FieldMetadata, requestValue: unknown, location: string): unknown {
		if (!Array.isArray(requestValue)) {
			if (field.options.fromWhere === 'query') {
				requestValue = [requestValue];
			} else {
				this.errors.push(new RequestStructureError(`Property ${field.name} is not an array`, location));
				return;
			}
		}

		const aux = requestValue as unknown[];
		const minLength = field.options.requirements?.minItems ?? 0;
		const maxLength = field.options.requirements?.maxItems ?? Number.MAX_SAFE_INTEGER;

		if (aux.length < minLength || aux.length > maxLength) {
			this.errors.push(
				new RequestStructureError(
					`${field.name} must have between ${minLength} and ${maxLength} items`,
					location,
				),
			);
			return;
		}

		if (!field.options.itemsExpects) {
			this.errors.push(
				new RequestStructureError(`Items expects not defined for property array ${field.name}`, location),
			);
			return;
		}

		if (field.options.itemsExpects === 'object' && !field.options.objFactory) {
			this.errors.push(
				new RequestStructureError(`typeFactory defined for property object ${field.name}`, location),
			);
			return;
		}

		let arrayIdx = 0;
		const output: unknown[] = [];

		for (const item of aux) {
			const currentLocation = `${location}${field.options.isBody ? '' : field.name}[${arrayIdx++}]`;
			if (field.options.itemsExpects === 'object' && field.innerFields) {
				const obj = getInstance(field.options.objFactory);
				output.push(obj);

				this.innerBind(currentLocation, obj as JsObj, field.innerFields.fields, item);
			} else {
				output.push(this.getPropertParsedValue(field, item, currentLocation, field.options.itemsExpects));
			}
		}

		return output;
	}

	private checkTargetConstructor(options: IInternalRequestBindOptions, parsedValue: ParseOutput, location: string) {
		if (options.targetConstructor === undefined) {
			return;
		}

		try {
			parsedValue.parsedValue = options.targetConstructor(parsedValue.parsedValue);
		} catch (err) {
			const error = err as Error;
			this.errors.push(new ValidationError(error.message, location, error.name));
		}
	}

	public hasErrors(): boolean {
		return true;
	}
}

function getInstance(factory: (() => unknown) | undefined): unknown {
	let aux: unknown = factory;
	while (typeof aux === 'function') {
		aux = aux();
	}

	return aux;
}
