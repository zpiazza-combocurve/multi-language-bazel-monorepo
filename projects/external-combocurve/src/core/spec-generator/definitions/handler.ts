/* eslint-disable 
	complexity,
	@typescript-eslint/explicit-module-boundary-types, 
	@typescript-eslint/no-explicit-any,
	@typescript-eslint/no-non-null-assertion */
import Handlebars from 'handlebars';

import { definitionsTemplate, ObjectDefinitionSpec } from '../templates';
import { IInternalRequestBindOptions, RequestBindOptionsKey, RequestBindPropertiesKey } from '../../common';
import { Controller } from '../../controllers/base';
import { fillDefinitionExamples } from '../example.helper';
import { PayloadCommandRequest } from '../../requests/base';
import { specs } from '../../metadata/metadata';

import { reqParseInput, requestSpecParses, resParseInput, responseSpecParses } from './parses';

type sampleObjContext = {
	obj: any; // The object sample
	name?: string; // The name of the definitions. It's different for reponse and inner objects
	parentKey?: string; // When it's a inner obj of a response, this is the key of the parent object
};

const statusNameMap: Record<number, string> = {
	200: 'OK',
	201: 'Created',
	202: 'Accepted',
	204: 'NoContent',
	207: 'MultiStatus',
	400: 'BadRequest',
	401: 'Unauthorized',
	403: 'Forbidden',
	404: 'NotFound',
};

export class DefinitionsGenHandler<TController extends Controller> {
	private controller: TController;
	private definitionsBasePath: string;

	constructor(controller: TController, definitionsBasePath?: string) {
		this.controller = controller;
		this.definitionsBasePath = definitionsBasePath ?? '#/definitions/';
	}

	public getDefinitionSpec(): string {
		const requestHash = new Set<string>();
		const objectsDefinitions: ObjectDefinitionSpec[] = [];

		for (const route of this.controller.Routes) {
			const routeRequestModel = route.requestSample;
			const requestName = this.getNameFromType(routeRequestModel);

			// Sometimes we can use the same request model for different routes
			if (requestHash.has(requestName)) {
				continue;
			}

			requestHash.add(requestName);

			objectsDefinitions.push(...this.mapRequestObjects(routeRequestModel));
			objectsDefinitions.push(...this.mapResponseSamples(requestName, routeRequestModel as Record<string, any>));
		}

		// Create the complete examples
		fillDefinitionExamples(objectsDefinitions);

		this.addDefinitionsBasePath(objectsDefinitions);

		const template = Handlebars.compile(definitionsTemplate);
		const result = template({
			objects: objectsDefinitions.filter((f) => f.properties.length > 0),
		});

		// The template has a empty line at the beginning, so we need to remove it
		return result.substring(result.indexOf('\n') + 1);
	}

	private addDefinitionsBasePath(objectsDefinitions: ObjectDefinitionSpec[]): void {
		for (const obj of objectsDefinitions) {
			for (const objProp of obj.properties.filter((p) => p.typeKey === '$ref')) {
				objProp.type = this.definitionsBasePath + objProp.type;
			}

			for (const objProp of obj.properties.filter((p) => p.itemsTypeKey === '$ref')) {
				objProp.type = this.definitionsBasePath + objProp.itemsType;
			}
		}
	}

	public mapRequestObjects(request: any): ObjectDefinitionSpec[] {
		// Run through request's properties and get all objects
		// This is pretty much a single graph traversal

		const output: ObjectDefinitionSpec[] = [];
		const objects: any[] = [request];

		while (objects.length > 0) {
			const currentObject = objects.pop();
			const objectSpec: ObjectDefinitionSpec = {
				name: this.getNameFromType(currentObject),
				properties: [],
				required: [],
			};

			const bindProperties = Reflect.getMetadata(RequestBindPropertiesKey, currentObject) as string[];

			for (const propertyKey of bindProperties ?? []) {
				const options = Reflect.getMetadata(
					RequestBindOptionsKey,
					currentObject,
					propertyKey,
				) as IInternalRequestBindOptions;

				// Spec definitions just consider body objects
				if (!options || (options.fromWhere !== 'body' && options.fromWhere !== 'scope')) {
					continue;
				}

				// CommandRequest has a special treatment, because it's have a payload property that should be ignored
				if (options.isBody && this.isCommandRequest(currentObject, options) && options.objFactory) {
					objects.push(options.objFactory());
					continue;
				}

				if (options.expects === 'object') {
					// We have a unique edge case that's when the propert is a object but have a sample.
					// In general this means the object does not have a bunch of these decorators
					// so we can't get the properties from metadata
					if (options.specOptions?.sample) {
						this.requestSampleEdgeCase(propertyKey, options, objectSpec, output);
						continue;
					}

					objects.push(options.objFactory!());
				}

				const specParse = requestSpecParses[options.expects];
				if (specParse) {
					const input: reqParseInput = {
						prop: propertyKey,
						opt: options,
					};

					objectSpec.properties.push(specParse(input));
				}

				if (options.isOptional !== true) {
					objectSpec.required!.push(propertyKey);
				}
			}

			output.push(objectSpec);
		}

		return output;
	}

	/** This method is responsible for getting all objects from a request sample */
	private requestSampleEdgeCase(
		propertyKey: string,
		bindingOptions: IInternalRequestBindOptions,
		currentObjSpec: ObjectDefinitionSpec,
		requestObjSpecs: ObjectDefinitionSpec[],
	) {
		// Get all definitions from this object sample
		const objDefinitions = this.getObjSpecFromSample(
			[{ obj: bindingOptions.specOptions?.sample, parentKey: propertyKey }],
			currentObjSpec.name,
			'Input',
		);

		// Then point the current request definition property to the first definition cretated by the sample
		if (objDefinitions.length > 0) {
			requestObjSpecs.push(...objDefinitions);
			currentObjSpec.properties.push({
				name: propertyKey,
				type: objDefinitions[0].name,
				typeKey: '$ref',
			});
		}
	}

	public mapResponseSamples(requestName: string, request: Record<string, any>): ObjectDefinitionSpec[] {
		const objs: sampleObjContext[] = [];

		// Run through all request's spec options registered on metadata
		for (const responseSpecOptions of specs.responseOptions(request.constructor).sort((s) => +s.status)) {
			if (responseSpecOptions.schema) {
				const statusName = statusNameMap[responseSpecOptions.status] ?? '';

				// The default definition name for the responses will be the request name + status name + Response
				// Example: GetUsersOKResponse, GetUsersNotFoundResponse
				objs.push({
					obj: responseSpecOptions.schema,
					name: `${requestName}${statusName}Response`,
				});
			}
		}

		return this.getObjSpecFromSample(objs, requestName, 'Output');
	}

	private getObjSpecFromSample(objs: sampleObjContext[], requestName: string, suffix: string) {
		// Run through all object entries, if found a new object add it to the stack
		// Based on the value type create the definition object properties

		const output: ObjectDefinitionSpec[] = [];
		while (objs.length > 0) {
			const currentObj = objs.pop()!;

			// The response's objects had their names defined based on status code + 'Response' suffix
			// The inner objects will have their names defined based on the RequestName + parent object property name + suffix
			// Example: GetUsersAddressListOutput, GetUsersZipCodeOutput
			const objectSpec: ObjectDefinitionSpec = {
				name: currentObj.name ?? `${requestName}${capitalize(currentObj.parentKey)}${suffix}`,
				properties: [],
			};

			Object.entries(currentObj.obj).forEach(([key, value]) => {
				if (value !== undefined) {
					let propType = typeof value as string;

					// If the value is an array, it's necessary to check if we have more objects inside it
					if (propType === 'object' && Array.isArray(value)) {
						propType = 'array';

						const firstItem = value[0];
						if (typeof firstItem === 'object') {
							objs.push({
								obj: firstItem,
								parentKey: key,
							});
						}
					}

					const parse = responseSpecParses[propType];
					const input: resParseInput = {
						prop: key,
						value: value,
						requestName: requestName,
						objSuffix: suffix,
					};

					objectSpec.properties.push(parse(input));

					if (propType === 'object' && !Array.isArray(value)) {
						objs.push({
							obj: value,
							parentKey: key,
						});
					}
				}
			});

			output.push(objectSpec);
		}

		return output;
	}

	private getNameFromType(type: any): string {
		return Object.getPrototypeOf(type).constructor.name;
	}

	private isCommandRequest(requestModel: unknown, options: IInternalRequestBindOptions): boolean {
		if (requestModel instanceof PayloadCommandRequest) {
			const commandReq = requestModel as PayloadCommandRequest<unknown, unknown>;
			options.objFactory = commandReq.payloadFactory;
			return true;
		}

		return false;
	}
}

function capitalize(s?: string): string {
	return s ? s.charAt(0).toUpperCase() + s.slice(1) : '';
}
