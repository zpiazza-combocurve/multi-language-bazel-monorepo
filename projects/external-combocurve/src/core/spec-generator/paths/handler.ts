/* eslint-disable 
	complexity,
	@typescript-eslint/explicit-module-boundary-types, 
	@typescript-eslint/no-explicit-any,
	@typescript-eslint/no-non-null-assertion */
import Handlebars from 'handlebars';

import { IInternalRequestBindOptions, RequestBindOptionsKey, RequestBindPropertiesKey } from '@src/core/common';
import { PayloadCommandRequest } from '@src/core/requests/base';
import { specs } from '@src/core/metadata/metadata';

import { Controller, RouteRequest } from '../../controllers/base';
import { EndpointParameterSpec, EndpointPathSpec, EndpointResponseSpec, pathsTemplate } from '../templates';
import { getExample } from '../example.helper';

const statusNameMap: Record<number, string> = {
	200: 'OK',
	201: 'Created',
	202: 'Accepted',
	204: 'NoContent',
	207: 'MultiStatus',
	400: 'BadRequest',
	401: 'Unauthorized',
	403: 'Forbidden',
	404: 'NotFound',
};

export class PathsGenHandler<TController extends Controller> {
	private controller: TController;
	private definitionsBasePath: string;
	private baseRoute: string;
	private idPrefix: string;

	constructor(baseRoute: string, controller: TController, idPrefix?: string, definitionsBasePath?: string) {
		this.baseRoute = baseRoute;
		this.controller = controller;
		this.idPrefix = idPrefix ?? 'v1';
		this.definitionsBasePath = definitionsBasePath ?? '#/definitions/';
	}

	public getPathSpec(): string {
		const output: Record<string, EndpointPathSpec[]> = {};

		for (const routeRegister of this.controller.Routes) {
			const routeEndpoint = this.baseRoute + routeRegister.route;
			if (routeRegister.notGenerateSpecs !== true) {
				const routeSpec = this.getRouteSpec(routeRegister);

				const methodsForRoute = output[routeEndpoint] ?? [];
				methodsForRoute.push(routeSpec);

				output[routeEndpoint] = methodsForRoute;
			}
		}

		const templateInput = {
			routes: Object.entries(output).map(([endpoint, methods]) => ({
				completeRoute: endpoint,
				methods,
			})),
		};

		const template = Handlebars.compile(pathsTemplate);
		const result = template(templateInput);

		return result.substring(result.indexOf('\n') + 1);
	}

	public getRouteSpec(route: RouteRequest): EndpointPathSpec {
		const requestName = getNameFromType(route.requestSample).toLowerCase();
		const routeSpec: EndpointPathSpec = {
			verb: route.method,
			id: `${route.method}-${this.idPrefix}-${requestName}`,
			parameters: [],
			responses: [],
		};

		if (route.method === 'post' || route.method === 'put') {
			routeSpec.consumes = 'application/json';
			routeSpec.produces = 'application/json';
		}

		routeSpec.parameters = [...this.getParameters(route.requestSample)];
		routeSpec.responses = [...this.getResponses(route.requestSample)];

		return routeSpec;
	}

	public *getParameters(requestSample: any): Generator<EndpointParameterSpec> {
		const bindProperties = Reflect.getMetadata(RequestBindPropertiesKey, requestSample) as string[];
		for (const propertyKey of bindProperties ?? []) {
			const options = Reflect.getMetadata(
				RequestBindOptionsKey,
				requestSample,
				propertyKey,
			) as IInternalRequestBindOptions;

			if (!options || options.fromWhere === 'services') {
				continue;
			}

			if (options.fromWhere === 'composition') {
				for (const param of this.getParameters(
					options.objFactory ? options.objFactory() : requestSample[propertyKey],
				)) {
					yield param;
				}

				continue;
			}

			if (options.expects === 'take' || options.expects === 'skip' || options.expects === 'sort') {
				yield this.getSearchParam(requestSample, options);
			}

			const parameter: any = {
				fromWhere: options.fromWhere,
				name: propertyKey,
				required: options.isOptional ? false : true,
				type: options.expects === 'objectID' ? 'string' : options.expects,
				example: getExample(options.expects, propertyKey),
			};

			if (!this.checkPayloadCommandRequestEdgeCase(requestSample, parameter, options)) {
				if (options.expects === 'object') {
					this.completeObjectParameter(propertyKey, requestSample, parameter, options);
				} else if (options.expects === 'array') {
					this.completeArrayParameter(parameter, options);
				}
			}

			yield parameter as EndpointParameterSpec;
		}
	}

	private getSearchParam(requestSample: any, options: IInternalRequestBindOptions): EndpointParameterSpec {
		const searchParam = options.expects as 'take' | 'skip' | 'sort';
		switch (searchParam) {
			case 'skip':
				return {
					fromWhere: 'query',
					name: 'skip',
					description: 'number of items to skip',
					default: requestSample.skip ?? '0',
					type: 'integer',
					format: 'int32',
				};
			case 'take':
				return {
					fromWhere: 'query',
					name: 'take',
					type: 'integer',
					format: 'int32',
					description: 'max records to return',
					default: requestSample.take ?? '25',
					minimum: 1,
					maximum: 200,
				};
			case 'sort':
				return {
					fromWhere: 'query',
					name: 'sort',
					type: 'string',
					default: 'id',
					description:
						'field to sort by, including + or - at the beginning for ascending or descending order, respectively',
					stringEnum: options.requirements?.validValues as string[],
				};
		}
	}

	private checkPayloadCommandRequestEdgeCase(
		requestSample: any,
		parameter: any,
		options: IInternalRequestBindOptions,
	): boolean {
		if (options.isBody && requestSample instanceof PayloadCommandRequest) {
			parameter.isArray = true;
			parameter.itemsTypeKey = '$ref';
			parameter.itemsType = `${this.definitionsBasePath}${getNameFromType(options.objFactory!())}`;
			return true;
		}

		return false;
	}

	private completeObjectParameter(
		propertyKey: string,
		requestSample: any,
		parameter: any,
		options: IInternalRequestBindOptions,
	) {
		delete parameter.type;

		if (options.customParse !== undefined && options.objFactory === undefined) {
			const reqName = getNameFromType(requestSample);
			const propName = capitalize(propertyKey);

			parameter.schema = `${this.definitionsBasePath}${reqName}${propName}Input`;
			return;
		}

		parameter.schema = `${this.definitionsBasePath}${getNameFromType(options.objFactory!())}`;
	}

	private completeArrayParameter(parameter: any, options: IInternalRequestBindOptions) {
		parameter.isArray = true;
		parameter.minItems = options.requirements?.minItems;
		parameter.maxItems = options.requirements?.maxItems;

		delete parameter.example;

		if (options.expects === 'object') {
			const arrayObj = options.objFactory!();

			parameter.itemsTypeKey = '$ref';
			parameter.itemsType = `${this.definitionsBasePath}${getNameFromType(arrayObj)}`;
		} else {
			parameter.itemsTypeKey = 'type';
			parameter.itemsType = options.itemsExpects === 'objectID' ? 'string' : options.itemsExpects;
			parameter.itemsExample = getExample(options.itemsExpects!, '');
		}
	}

	public *getResponses(requestSample: any): Generator<EndpointResponseSpec> {
		for (const responseSpecOptions of specs.responseOptions(requestSample.constructor).sort((s) => +s.status)) {
			const statusName = statusNameMap[responseSpecOptions.status] ?? '';

			const responseSpec: EndpointResponseSpec = {
				status: responseSpecOptions.status,
				description: `${statusName} Response`,
			};

			if (responseSpecOptions.schema) {
				const requestName = getNameFromType(requestSample);
				responseSpec.schema = `${this.definitionsBasePath}${requestName}${statusName}Response`;
			}

			if (responseSpecOptions.headers) {
				responseSpec.headers = Object.entries(responseSpecOptions.headers).map(([key, value]) => ({
					name: key,
					type: typeof value,
				}));
			}

			yield responseSpec;
		}
	}
}

function getNameFromType(type: any): string {
	return Object.getPrototypeOf(type).constructor.name;
}

function capitalize(s?: string): string {
	return s ? s.charAt(0).toUpperCase() + s.slice(1) : '';
}
