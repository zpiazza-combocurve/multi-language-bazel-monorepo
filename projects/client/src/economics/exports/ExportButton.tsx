import { omit } from 'lodash';

import { useDownloadFile } from '@/economics/Economics/shared/helpers';
import { LastRunSummaryQuery } from '@/economics/shared/queries';
import { isGHGReport, isPDFReport } from '@/economics/shared/shared';
import useLDFeatureFlags from '@/feature-flags/useLDFeatureFlags';
import { withLoadingBar } from '@/helpers/alerts';

import { CSVExportTemplateBase } from './CSVExportDialog/types';
import { downloadReport } from './ExportButton/helpers';
import { AllExportMenu } from './ExportButton/menus/AllExportMenu';
import { CSVExportMenu } from './ExportButton/menus/CSVExportMenu';
import { PDFExportMenu } from './ExportButton/menus/PDFExportMenu';
import { PDFExportTemplateBase } from './PDFExportDialog/shared/types';

export function ExportButton({
	runId,
	hasReservesGroups,
	scenarioTableHeaders,
	runningEconomics,
	ghgRunId,
	scenarioId,
	hasOneLiner,
}) {
	const [fileNameDialog, promptDownload] = useDownloadFile();

	const { isCustomPDFEditorEnabled } = useLDFeatureFlags();

	async function handleExport(props: CSVExportTemplateBase | PDFExportTemplateBase) {
		const template = omit(props, 'name');
		const isGHGRun = isGHGReport(template.type);
		const run = isGHGRun ? ghgRunId : runId;

		const result = await promptDownload(template);

		if (!result) return;

		const { fileName, bfitReport, afitReport } = result;

		const promise = downloadReport({
			...template,
			fileName,
			run,
			bfitReport,
			afitReport,
		});

		if (isPDFReport(template.type)) {
			// skip loading bar for pdf // why though?
			await promise;
		} else {
			await withLoadingBar(promise);
		}

		if (template.cashflowOptions?.type === 'monthly') {
			LastRunSummaryQuery.invalidate(scenarioId);
		}
	}

	const sharedMenuProps = {
		runId,
		runningEconomics,
		handleExport,
		scenarioTableHeaders,
		econRunExists: !!runId,
		ghgRunExists: !!ghgRunId,
		hasReservesGroups,
		hasOneLiner,
	};

	return (
		<>
			{fileNameDialog}
			{isCustomPDFEditorEnabled ? (
				<>
					<CSVExportMenu {...sharedMenuProps} />
					<PDFExportMenu {...sharedMenuProps} />
				</>
			) : (
				<AllExportMenu {...sharedMenuProps} />
			)}
		</>
	);
}
