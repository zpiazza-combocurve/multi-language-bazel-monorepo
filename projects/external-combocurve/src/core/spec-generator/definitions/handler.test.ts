import { TestController, TestModel } from '@src/core/test.exports';

import { DefinitionsGenHandler } from './handler';

describe('spec handler', () => {
	it('mapRequestObjects', () => {
		const model = new TestModel();
		const handler = new DefinitionsGenHandler<TestController>(new TestController());

		const definitions = handler.mapRequestObjects(model);

		expect(definitions.length).toBe(3);

		const customInputModel = definitions[0];
		const testModel = definitions[1];
		const innerModel = definitions[2];

		// Name
		expect(testModel.name).toBe('TestModel');
		expect(innerModel.name).toBe('InnerModel');
		expect(customInputModel.name).toBe('TestModelCustomInput');

		// Required
		expect(testModel.required).toEqual(['name', 'isCool', 'inner']);
		expect(innerModel.required).toEqual(['profession', 'yearsOfExperience', 'lookingForJob']);

		// Properties
		// TODO
	});

	it('mapResponseObjects', () => {
		const model = new TestModel();
		const handler = new DefinitionsGenHandler<TestController>(new TestController());

		const definitions = handler.mapResponseSamples('TestModel', model);

		expect(definitions.length).toBe(5);

		const ok = definitions[0];
		const innerOutput = definitions[1];
		const notFound = definitions[2];
		const created = definitions[3];
		const peoplesOutput = definitions[4];

		// Name
		expect(created.name).toBe('TestModelCreatedResponse');
		expect(peoplesOutput.name).toBe('TestModelPeoplesOutput');
		expect(notFound.name).toBe('TestModelNotFoundResponse');
		expect(ok.name).toBe('TestModelOKResponse');
		expect(innerOutput.name).toBe('TestModelInnerOutput');

		// Properties
		expect(created.properties.length).toBe(1);
		expect(peoplesOutput.properties.length).toBe(2);
		expect(notFound.properties.length).toBe(3);
		expect(ok.properties.length).toBe(7);
		expect(innerOutput.properties.length).toBe(3);
	});

	it('getDefinitions', () => {
		const handler = new DefinitionsGenHandler<TestController>(new TestController());

		const yaml = handler.getDefinitionSpec();

		expect(yaml).toMatchSnapshot();
	});
});
