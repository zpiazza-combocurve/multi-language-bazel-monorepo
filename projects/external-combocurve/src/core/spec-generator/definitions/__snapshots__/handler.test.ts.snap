// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec handler getDefinitions 1`] = `
"TestModelCustomInput:
  type: 'object'
  properties:
    prop1:
      type: 'string'
      example: 'qwe'
    prop2:
      type: 'number'
      example: '123'
  example:
    prop1: qwe	
    prop2: 123	
TestModel:
  type: 'object'
  required:
    - name
    - isCool
    - inner
  properties:
    name:
      type: 'string'
      example: 'test1'
      enum:
        - 'test1'
        - 'test2'
    isCool:
      type: 'boolean'
      example: 'true'
    inner:
      $ref: '#/definitions/InnerModel'
    address:
      type: 'string'
      example: 'some_string'
    custom:
      $ref: '#/definitions/TestModelCustomInput'
  example:
    name: test1	
    isCool: true	
    inner:	
      profession: some_string	
      yearsOfExperience: 42	
      lookingForJob: true	
    address: some_string	
    custom:	
      prop1: qwe	
      prop2: 123	
InnerModel:
  type: 'object'
  required:
    - profession
    - yearsOfExperience
    - lookingForJob
  properties:
    profession:
      type: 'string'
      example: 'some_string'
    yearsOfExperience:
      type: 'number'
      format: 'int32'
      example: '42'
    lookingForJob:
      type: 'boolean'
      example: 'true'
  example:
    profession: some_string	
    yearsOfExperience: 42	
    lookingForJob: true	
TestModelOKResponse:
  type: 'object'
  properties:
    name:
      type: 'string'
      example: 'test'
    age:
      type: 'number'
      example: '10'
    isCool:
      type: 'boolean'
      example: 'true'
    id:
      type: 'string'
      example: '5f0b4e3b9b0b3e1b3c9d3b3b'
    inner:
      $ref: '#/definitions/TestModelInnerOutput'
    address:
      type: 'string'
      example: 'default address'
    numbers:
      type: 'array'
      items:
        type: 'string'
        example: 'one'
  example:
    name: test	
    age: 10	
    isCool: true	
    id: 5f0b4e3b9b0b3e1b3c9d3b3b	
    inner:	
      profession: developer	
      yearsOfExperience: 10	
      lookingForJob: false	
    address: default address	
    numbers:	
      - one	
TestModelInnerOutput:
  type: 'object'
  properties:
    profession:
      type: 'string'
      example: 'developer'
    yearsOfExperience:
      type: 'number'
      example: '10'
    lookingForJob:
      type: 'boolean'
      example: 'false'
  example:
    profession: developer	
    yearsOfExperience: 10	
    lookingForJob: false	
TestModelNotFoundResponse:
  type: 'object'
  properties:
    name:
      type: 'string'
      example: 'test'
    errorMsg:
      type: 'string'
      example: 'test'
    location:
      type: 'string'
      example: 'test'
  example:
    name: test	
    errorMsg: test	
    location: test	
TestModelCreatedResponse:
  type: 'object'
  properties:
    peoples:
      type: 'array'
      items:
        $ref: 'TestModelPeoplesOutput'
  example:
    peoples:	
      - name: test	
        age: 10	
TestModelPeoplesOutput:
  type: 'object'
  properties:
    name:
      type: 'string'
      example: 'test'
    age:
      type: 'number'
      example: '10'
  example:
    name: test	
    age: 10	
"
`;
